### PR TITLE
Implement methods to list all components and all resources (Rebase of PR #3012)

### PR DESCRIPTION
**Pull request by KevinKeyser**
# Objective

- Take over and Rebase Pull Request https://github.com/bevyengine/bevy/pull/3012
    - Fixes https://github.com/bevyengine/bevy/issues/3007
    - Simplify debugging by allowing iteration over all Components and Resources registered in a World.

## Solution

- [x] Rename Components into WorldData to reflect that it stores metadata both on Components and Resources. Also rename ComponentId to DataId, ComponentInfo to DataInfo and ComponentDescriptor to DataDescriptor
- [x] implement IntoIterator for WorldData to iterate over all components and resources
- [x] create a new Components system descriptor and implement IntoIterator to iterate over all components
- [x] create a new Resources system descriptor and implement IntoIterator to iterate over all resources

---

## Changelog

- `struct ComponentId` -> `struct DataId`
- `struct Components` -> `struct WorldData`
   - `Components::components` -> `WorldData::data`
- `struct ComponentInfo` -> `struct DataInfo`
- `struct ComponentDescriptor` -> `struct DataDescriptor`

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
My instinct is that we should merge #4809 first, which should simplify the implementation here significantly. @james7132, are you in agreement?

**Comment by targrub** [comment]
@KevinKeyser Is this a dead PR at this point?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I like the `ComponentId` -> `DataId` change: I think it's genuinely correct and clearer. However it makes this PR much harder to review, and should be its own PR.

File: crates/bevy_ecs/src/system/mod.rs [hunk 1]
```diff
@@ -517,7 +519,49 @@ mod tests {
...
+
+        run_system(&mut world, sys);
+        assert_eq!(world.get_resource::<Vec<String>>().unwrap().len(), 3);
+    }
+
+    #[test]
+    fn iterate_resources() {
+        let mut world = World::default();
+        world.insert_resource(Vec::<String>::new());
+        world.insert_resource(42usize);
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/mod.rs:544]: These will have to be redone to account for the opt-in `Resource` trait.

File: crates/bevy_ecs/src/system/mod.rs [hunk 2]
```diff
@@ -517,7 +519,49 @@ mod tests {
...
+
+        run_system(&mut world, sys);
+        assert!(world
+            .get_resource::<Vec<String>>()
+            .unwrap()
+            .contains(&"usize".to_string()),);
+    }
+
+    #[test]
+    fn remove_tracking() {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/mod.rs:564]:
```suggestion
    fn removal_tracking() {
```

Just to reduce diff noise.

**Review by alice-i-cecile — APPROVED** [review]
This is quite nice: I really like how much clearer `DataId` and `WorldData` are. If you rebase this and get CI passing I'd be happy to see this merged basically as is.

**Comment by alice-i-cecile** [comment]
@KevinKeyser I still really like this work, but at this point it's going to be easier to remake this PR than to solve merge conflicts. I'm going to `Adopt-Me` to this PR, but feel free to adopt it yourself if you'd like!

**Comment by Adamkob12** [comment]
I'd like to pick this up soon.

File: crates/bevy_ecs/src/system/system_param.rs [hunk 3]
```diff
@@ -1107,26 +1107,131 @@ impl<'w, 's> SystemParamFetch<'w, 's> for ArchetypesState {
...
+    type Fetch = ComponentsState;
+}
 
 /// The [`SystemParamState`] of [`Components`].
 #[doc(hidden)]
 pub struct ComponentsState;
 
 // SAFE: no component value access
-unsafe impl SystemParamState for ComponentsState {
+unsafe impl<'a> SystemParamState for ComponentsState {
```
Comment (AustinHellerRepo) [inline crates/bevy_ecs/src/system/system_param.rs:1168]: Probably trivial, but is this explicit lifetime required?

File: crates/bevy_ecs/src/world/mod.rs [hunk 4]
```diff
@@ -697,22 +692,22 @@ impl World {
...
     pub fn contains_resource<R: 'static>(&self) -> bool {
-        let component_id =
-            if let Some(component_id) = self.components.get_resource_id(TypeId::of::<R>()) {
-                component_id
-            } else {
-                return false;
-            };
+        let component_id = if let Some(component_id) = self.data.get_resource_id(TypeId::of::<R>())
+        {
+            component_id
```
Comment (AustinHellerRepo) [inline crates/bevy_ecs/src/world/mod.rs:697]: Should this be `resource_id`?

**Comment by AustinHellerRepo** [comment]
I'm hoping that this PR will help permit me to more easily serialize specific components (when I only have the name) of an entity as well as overwrite the state of a component by deserializing a previous state's `String` into the component.
If this is now possible with these changes, would it be possible to get a unit test or example included in this PR?
If that is out of scope for these changes, any help with this task would be greatly appreciated. Thank you.

**Comment by alice-i-cecile** [comment]
@matiqo15, this seems really useful for https://github.com/bevyengine/rfcs/pull/77. Perhaps we should re-adopt it and get the basic form merged?

**Comment by matiqo15** [comment]
> @matiqo15, this seems really useful for https://github.com/bevyengine/rfcs/pull/77. Perhaps we should re-adopt it and get the basic form merged?

Agree, this could help a lot

**Comment by VitalyAnkh** [comment]
@alice-i-cecile Now that #12829  has been merged, is this PR still necessary?

**Comment by alice-i-cecile** [comment]
It would still be nice to have something like this for components, but I'm going to close this out regardless.

**Pull request closed without merging**